internal/storage/inmem/query.go

```diff
@@ -0,0 +1,187 @@
+package inmem
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// indexSeparator delimits the segments of our radix tree keys.
+const indexSeparator = "\x00"
+
+// readIndexer implements the memdb.Indexer interface (see FromArgs).
+type readIndexer struct{}
+
+// FromArgs constructs a radix tree key from an ID for lookup.
+func (i readIndexer) FromArgs(args ...any) ([]byte, error) {
+	if l := len(args); l != 1 {
+		return nil, fmt.Errorf("expected 1 arg, got: %d", l)
+	}
+	id, ok := args[0].(*pbresource.ID)
+	if !ok {
+		return nil, fmt.Errorf("expected *pbresource.ID, got: %T", args[0])
+	}
+	return indexFromID(id), nil
+}
+
+// idIndexer implements the memdb.SingleIndexer and memdb.PrefixIndexer
+// interfaces. It is used for indexing resources by their IDs.
+type idIndexer struct{ readIndexer }
+
+// FromObject constructs a radix tree key from a Resource at write-time, or an
+// ID at delete-time.
+func (i idIndexer) FromObject(raw any) (bool, []byte, error) {
+	switch t := raw.(type) {
+	case *pbresource.ID:
+		return true, indexFromID(t), nil
+	case *pbresource.Resource:
+		return true, indexFromID(t.Id), nil
+	}
+	return false, nil, fmt.Errorf("expected *pbresource.Resource or *pbresource.ID, got: %T", raw)
+}
+
+// PrefixFromArgs constructs a radix tree key prefix from a query for listing.
+func (i idIndexer) PrefixFromArgs(args ...any) ([]byte, error) {
+	if l := len(args); l != 1 {
+		return nil, fmt.Errorf("expected 1 arg, got: %d", l)
+	}
+
+	q, ok := args[0].(query)
+	if !ok {
+		return nil, fmt.Errorf("expected query, got: %T", args[0])
+	}
+	return q.indexPrefix(), nil
+}
+
+// ownerIndexer implements the memdb.SingleIndexer interface. It is used for
+// indexing resources by their owners.
+type ownerIndexer struct{ readIndexer }
```

Comment:
Just adding this here as an eventual idea, but if we're going to use this eventually as a local cache for remote backends, we may consider exposing these indexers to be extended on a per-type basis. The corresponding local indexers for Kubernetes allow you to do something similar, basically letting you index some arbitrary fields via a call to [IndexField](https://github.com/kubernetes-sigs/controller-runtime/blob/e2d8821b277fb57f3a951f430bb1746d8297b853/pkg/cache/informer_cache.go#L161-L166) in your setup code and then query them via an explicit `FieldSelector` parameter on your query.

See envoy's gateway controller implementation for usage around calling [IndexField](https://github.com/envoyproxy/gateway/blob/18890ecc7e526cfb5d1ee5958237badff215a12b/internal/provider/kubernetes/controller.go#L486-L488), an [example indexer](https://github.com/envoyproxy/gateway/blob/18890ecc7e526cfb5d1ee5958237badff215a12b/internal/provider/kubernetes/controller.go#L504-L523), and how the index is [being queried](https://github.com/envoyproxy/gateway/blob/main/internal/provider/kubernetes/routes.go#L178-L180). I'd imagine we'd eventually greatly benefit if we were able to do something like be able to add additional resource-based indices at resource registration time.

Reply:
Thanks for the idea, @andrewstucki! I've wondered about something similar too. I wonder if it'd be better to maintain a clean separation and have controllers implement their own MemDB stores (with their own custom indices) though?

Reply:
Yeah, I'm fine either way, and again, this is definitely not an "initial pass" sort of thing. It makes sense to me to have controller-specific stores/indices in the sense that you don't wind up having to worry as much about having potentially conflicting indexed fields -- that said, it would mean having more layers of caching/storage, alternatively we could eventually expose doing something like prefixing controller names into the indexed field and implicitly appending those prefixes to a lookup from the controller, that way they're essentially still acting on the monolithic cache and we have less memory usage -- but 🤷, we can figure that out in subsequent PRs/layers of the architecture.